### ISSUE TITLE
Make updatability of package.json clear

### ISSUE THREAD

**Issue by Legioth** [vaadin/flow#5465]
Right now, the way `package.json` is managed is really unclear for users.

The good part with the current approach makes it possible for users to add new custom dependencies unrelated to their Java code (e.g. `sass-loader` for webpack). Where it falls apart is any of these cases:
* If the user updates the version of a dependency in their `pom.xml`, then the corresponding dependency in `package.json` remains on its old version. This can to some degree be fixed by adding logic that explicitly forces the version in `package.json` to be in sync with whatever is found from annotations on the classpath. The drawback of that "fix"  is that the user will no longer be able to explicitly override to use an older or newer version of the JS module to e.g. work around some specific bug in the current version.
* If the user removes a dependency from their `pom.xml`, then the corresponding dependency is not removed from `package.json`. We cannot remove any dependencies from `package.json` without risking to remove something that the user has explicitly put there independently of what they happen to have on the classpath.

Both problems can trivially be solved by always recreating the file and thus never allowing the user to make any modifications.

If we want to allow the user to modify the dependencies, we should do so in a way that
* doesn't cause problems to those that haven't done any customization
* allows a user who wants to modify the file to actually be in control over the file

I see two basic approaches for achieving this. The simple approach is all-or-nothing: the user can choose between giving full control to Flow or prevent Flow from making any changes at all. This could be expressed e.g. as an option in a configuration file or by opting out if `package.json` contains any other top-level properties than `dependencies`  and `devDependencies`.

There is also a slightly more complex alternative that would make everything work automatically for users who don't want to make any own customization, but also make it possible to customize things while still benefitting from automatic resource management for anything referenced by Flow. The idea here would be to have a secondary `package.json` file that is always rewritten based on dependencies from Java classes. The primary `package.json` file would then be almost fully in the user's control, except that it would contain one entry that references the definition managed by Flow.

To summarize: we should either make the file explicitly out-of-bounds, or alternatively make it modifiable in a way that doesn't cause drawbacks for those who don't modify it. The current in-between approach is bad for both user categories.

**Comment by manolo** [comment]
> To summarize: we should either:

> make the file explicitly out-of-bounds

When developer wants its own custom file not being modifiable he can disable by removing the `update-npm-packages` goal in `pom.xml` (note that when we have one unique goal it will be a config parameter) 

> or alternatively make it modifiable in a way that doesn't cause drawbacks for those who don't modify it. 

 1. User can remove it from SCM, hence we create it each time they run a `git clean` 
 2. Because we do not add version field, it does not need to be updated, otherwise if user adds it, it's their responsability to keep it updated. I don't think version in `pom.xml` needs to be the same than in `package.json`. I neither think it's needed in since `package.json` file is here needed for installing dependencies but not for deploying anything to npm.


Another question is whether we update the version of each specific dependency, but that will need:
   - a way to mark those in the `@NpmPackage` annotation
   - or just use the magic `latest`, which means that user can still replace `latest` with a specific version, and we don't modify it anymore

Thus, If we decide to add a parameter to `@NpmPackage` API, we can consider these cases
  - `@NpmPackage(value = "@vaadin/vaadin-grid")` we set `"@vaadin/vaadin-grid": "latest"` and we do not modify never.
  - `@NpmPackage(value = "@vaadin/vaadin-grid", version="^5.3.3")` we always keep this  in sync   `"@vaadin/vaadin-grid": "^5.3.3"` if user wants to increase the version he needs to
   - update `vaadin-grid-flow` in `pom.xml` which will come with a new value in the annotation
   - annotate their view class with `@NpmPackage(value = "@vaadin/vaadin-grid", version="^6.0.0")` which overrides any other dependency found in their discovered components tree

**Comment by Legioth** [comment]
No need to keep the version of `package.json` itself in sync with `pom.xml`. I was only thinking of dependency versions. I've clarified that in the original description.

Not having any version number in `@NpmPackage` is a very serious omission that I wasn't even thinking of when writing the original descriptions. Without a way of specifying a specific version, things will be seriously broken for e.g. Grid which has so far turned out to be quite sensitive to exactly which version of the Web Component it's used together with. For other components, it would at the very least be necessary to define which major version to use, but for our own components it still best to keep having everything fully locked down so that builds can be reproducible (like we have to day also by defining exact versions for the webjars).

Let's imagine that between Vaadin 14 and Vaadin 15, both Charts and Grid get new major versions that have breaking changes.
* If you update your `pom.xml` from Vaadin 14 to 15, then you will get the Vaadin 15 version of the Java classes for Grid, but your `package.json` will still have a `^x.y.z` definition corresponding to the Vaadin 14 version of the web component. 💥 
* If you keep using Vaadin 14, but add `vaadin-charts` to your `pom.xml`, then the opposite will happen. `vaadin-bom` will give you the Vaadin 14 version of the Java classes, but `npm i` will give you the Vaadin 15 version of `@vaadin/vaadin-charts` since that is the latest one. 💥

**Comment by Legioth** [comment]
> User can remove it from SCM, hence we create it each time they run a `git clean`

Doing `git clean` every time you've done any changes to `pom.xml` sounds like a quite inconvenient way.

> own custom file not being modifiable he can disable by removing the update-npm-packages goal in pom.xml

Documenting this would be an improvement, but it would still only solve one of the problems. 

Remaining problems are:
* Removal from `package.json` after something has been removed from `pom.xml`
* The moment the developer disables this feature, they would have to manually update their `package.json` every time they update to a newer Vaadin version or add/remove any `pom.xml` dependency. We can decide that this is not important for now, but we won't get away with that in the long run.

**Comment by pleku** [comment]
Currently we also have an issue with getting "random failure" for a project because the dependencies might be broken and we don't know it, it depends on what `npm install` does: https://github.com/vaadin/flow/issues/5320

To my limited understanding it is because
- the lack of a package-lock file
- lack of the versioning information in the annotation ^ as claimed above

The original idea we had was to provide the version in the `package.json` for the component, but then we opted for the annotation instead. But I guess the version was removed (or not added) because we figured that webpack can figure out a correct version - but this can cause issues as presented ^.

So trying to summarize the problems please correct me if I'm wrong:
- we might have broken frontend dependencies at any point for the starters
- when updating the Vaadin version in pom.xml (where everyone will do it), it is not reflected on `package.json` and you don't get the fixes automatically. This is the same for any version bump.

**Comment by Legioth** [comment]
@pleku That is one part of the problem.

The second part is whether we want to support scenarios where the application developer wants to customize the dependencies. If decide that we don't want to support that (for now), then we should remove the current ambiguity that might give the developer the impression that it's actually supported. If we want to support it, then we should make it work in a meaningful way.

**Comment by pleku** [comment]
Target now:

_When updating the platform version, I expect to get the versions of the frontend dependencies updated automatically, without having to update anything myself because I would hate that._

_When creating an add-on, I want to declare the version for the npm module my add-on is based on, so that I can guarantee the users get a working version of the add-on when they take it into use._

Secondary:

_When building a project with the platform, I want to take control over certain frontend dependencies that are used, so that I can override stuff that I need to._

**Comment by caalador** [comment]
**Acceptance criteria:**

[ ] `NpmPackage`  should require the `version` to be given
[ ] `update-npm-dependencies` goal should collect versions from NpmPackage and write them to `package.json` by always overriding existing depedencies
[ ] Added dependencies should always be sorted for consistency

**Comment by mrts** [comment]
I think it would help to make it more visible and clear in documentation that the recommended (or only?) way to add packages to `package.json` is through the `@NpmPackage` annotation on Java classes.
